File: Circular_Linked_list/CircularDoublyLinkedListInsertOperations.java
// File Name: CircularDoublyLinkedListInsertOperations.java
// Program to insert a node at beginning, end, and given position
// in a circular doubly linked list

class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
        this.data = data;
    }
}

public class CircularDoublyLinkedListInsertOperations {
    static Node head = null;

    // Insert at beginning
    static void insertBeg(int value) {
        Node newNode = new Node(value);

        if (head == null) {
            newNode.next = newNode;
            newNode.prev = newNode;
            head = newNode;
            return;
        }

        Node last = head.prev;

        newNode.next = head;
        newNode.prev = last;
        last.next = newNode;
        head.prev = newNode;
        head = newNode;
    }

    // Insert at end
    static void insertEnd(int value) {
        Node newNode = new Node(value);

        if (head == null) {
            newNode.next = newNode;
            newNode.prev = newNode;
            head = newNode;
            return;
        }

        Node last = head.prev;

        newNode.next = head;
        newNode.prev = last;
        last.next = newNode;
        head.prev = newNode;
    }

    // Insert at given position (1-based)
    static void insertPos(int value, int pos) {
        if (head == null || pos == 1) {
            insertBeg(value);
            return;
        }

        Node temp = head;
        for (int i = 1; i < pos - 1 && temp.next != head; i++)
            temp = temp.next;

        Node newNode = new Node(value);
        newNode.next = temp.next;
        newNode.prev = temp;
        temp.next.prev = newNode;
        temp.next = newNode;
    }

    public static void main(String[] args) {
        insertBeg(10);
        insertEnd(30);
        insertPos(20, 2);
    }
}

File: Circular_Linked_list/circular_doubly_linked_list_insert_operations.c
// File Name: circular_doubly_linked_list_insert_operations.c
// Program to insert a node at beginning, end, and given position
// in a circular doubly linked list

#include <stdio.h>
#include <stdlib.h>

struct node {
    int data;
    struct node *prev;
    struct node *next;
};

struct node *head = NULL;

// Insert at beginning
void insert_beg(int value) {
    struct node *newnode = (struct node *)malloc(sizeof(struct node));
    newnode->data = value;

    if (head == NULL) {
        newnode->next = newnode;
        newnode->prev = newnode;
        head = newnode;
        return;
    }

    struct node *last = head->prev;

    newnode->next = head;
    newnode->prev = last;
    last->next = newnode;
    head->prev = newnode;
    head = newnode;
}

// Insert at end
void insert_end(int value) {
    struct node *newnode = (struct node *)malloc(sizeof(struct node));
    newnode->data = value;

    if (head == NULL) {
        newnode->next = newnode;
        newnode->prev = newnode;
        head = newnode;
        return;
    }

    struct node *last = head->prev;

    newnode->next = head;
    newnode->prev = last;
    last->next = newnode;
    head->prev = newnode;
}

// Insert at given position (1-based)
void insert_pos(int value, int pos) {
    if (head == NULL || pos == 1) {
        insert_beg(value);
        return;
    }

    struct node *temp = head;
    int i;

    for (i = 1; i < pos - 1 && temp->next != head; i++)
        temp = temp->next;

    struct node *newnode = (struct node *)malloc(sizeof(struct node));
    newnode->data = value;

    newnode->next = temp->next;
    newnode->prev = temp;
    temp->next->prev = newnode;
    temp->next = newnode;
}

int main() {
    insert_beg(10);
    insert_end(30);
    insert_pos(20, 2);
    return 0;
}
